src/VisualStudio/Core/Def/Implementation/Experimentation/KeybindingResetDetector.cs

```diff
@@ -0,0 +1,358 @@
+                    pCmdTrgt: this,
+                    pdwCookie: out _priorityCommandTargetCookie);
+
+                // Initialize the OleComponent to listen for modal changes (which will tell us when Tools->Options is closed)
+                _oleComponent = OleComponent.CreateHostedComponent("Keybinding Reset Detector");
+                _oleComponent.ModalStateChanged += OnModalStateChanged;
+
+                if (ErrorHandler.Failed(hr))
+                {
+                    FatalError.ReportWithoutCrash(Marshal.GetExceptionForHR(hr));
+                }
+            }
+        }
+
+        private void UpdateReSharperEnableStatus()
+        {
+            AssertIsForeground();
+
+            var isEnabled = IsReSharperEnabled();
+            var options = _workspace.Options;
+
+            if (isEnabled)
+            {
+                // Update the option if we weren't already enabled. There's no other actions to take, since
+                // ReSharper is enabled and we don't want to pop a gold bar
+                if (options.GetOption(KeybindingResetOptions.ReSharperStatus) != ReSharperStatus.Enabled)
+                {
+                    options = options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.Enabled)
+                                     .WithChangedOption(KeybindingResetOptions.NeedsReset, false);
+                    _workspace.Options = options;
+                }
+            }
+            else
+            {
+                bool needsReset;
+                if (options.GetOption(KeybindingResetOptions.ReSharperStatus) != ReSharperStatus.Suspended)
+                {
+                    options = options.WithChangedOption(KeybindingResetOptions.ReSharperStatus, ReSharperStatus.Suspended)
+                                     .WithChangedOption(KeybindingResetOptions.NeedsReset, true);
+                    needsReset = true;
+                    _workspace.Options = options;
+                }
+                else
+                {
+                    needsReset = options.GetOption(KeybindingResetOptions.NeedsReset);
+                }
+
+                if (needsReset)
+                {
+                    ShowGoldBar();
+                }
+            }
+        }
+
+        private bool IsReSharperEnabled()
+        {
+            AssertIsForeground();
+
+            if (_oleCommandTarget == null)
+            {
+                var oleServiceProvider = (IOleServiceProvider)_serviceProvider.GetService(typeof(IOleServiceProvider));
+                _oleCommandTarget = (IOleCommandTarget)oleServiceProvider.QueryService(VSConstants.SID_SUIHostCommandDispatcher);
+            }
+
+            var cmds = new OLECMD[1];
+            cmds[0].cmdID = SuspendId;
+            cmds[0].cmdf = 0;
+
+            ErrorHandler.ThrowOnFailure(_oleCommandTarget.QueryStatus(ReSharperCommandGroup, (uint)cmds.Length, cmds, IntPtr.Zero));
```

Comment:
will it cause resharper dll to be loaded to VS if resharper is installed but not loaded to Vs yet? or that doesn't matter?

Reply:
ReSharper will already be loaded by this point, we've already confirmed that the package is installed, and since we're running after the solution has loaded their hooks will be in place.

Reply:
Technically, there is no guarantee that resharper will be loaded at this point -- our package could load before solution load and before their trigger.

I don't believe that technicality matters.

Reply:
This is initialized after the solution has finished loading.